Package macOS releases as a universal binary.

pchote opened the pull request:
This PR adds native arm64 support for our packaged releases. .NET 6 does not support proper universal binaries, but our existing launcher structure made it relatively easy to emulate it by shipping both runtimes in the app and selecting the correct one at runtime.

One nice side-effect of this structure is that it allows us to treat mono as a third pseudo-architecture, so we can support macOS 10.11 - 10.14 without needing the separate -mono.dmg release. I made the choice to drop 10.9 and 10.10 support because this allows us to keep the better dmg compression added in 10.11, with the sysinfo database showing only 7 impacted players from the last 6 months.

Each runtime is deployed in its own directory, but the default net6 apphost (the native `OpenRA` executable created by the `publish` command) only works if it is located alongside the dylib and dll files. This doesn't play nicely with the macOS windowing system, which requires executables to live in the parent `MacOS` directory for the dock icon / app name associations to work.

These limitations are avoided by [creating our own minimal .NET host](https://learn.microsoft.com/en-gb/dotnet/core/tutorials/netcore-hosting) that is able to load the runtime files from a subdirectory. There's a bunch of discussion about improving the default .NET app host in a future release, so we may eventually be able to simplify this.

The `apphost-arm64` host is compiled as arm64-only, while the `apphost-x86_64` and `apphost-mono` hosts are compiled as intel-only. The Launcher entrypoint is compiled with both arm64 and x86_64 slices. The launcher checks the OS version, `OPENRA_PREFER_MONO` environment variable, and process architecture to decide which host to run.

The "Get Info -> Open using Rosetta" checkbox works as expected, using the x86_64 runtime through Rosetta. The mono runtime is used on macOS 10.11 - 10.14 or if the game is launched through the commandline as e.g. `OPENRA_PREFER_MONO=1 /Volumes/OpenRA/OpenRA\ -\ Dune\ 2000.app/Contents/MacOS/Launcher`.

The first commit is a bodge that works around the missing native binaries in the current nuget packages. If @teinarss can please package the libraries in the `deps` subdirectory we can rewrite that commit to use the updated packages for both packaged and development builds. They are generated with https://github.com/pchote/SDL2-binaries/commits/universal which also updates all platforms to SDL 2.26.1 and macOS to Freetype 2.12.1 and OpenAL-soft 1.22.2. I suggest that we rename the repo to `native-dependencies`, transfer it to the OpenRA organization, and add the rest of the missing windows/linux libraries as a future step.

An unsigned test build can be downloaded from https://github.com/pchote/OpenRA/releases/tag/devtest-20221223

pchote committed:
Report CPU arch in logs and sysinfo.

abcdefg30 commented on packaging/macos/apphost-mono.c:
```diff
@@ -0,0 +1,45 @@
+	void *libmono = dlopen(argv[1], RTLD_LAZY);
+	if (libmono == NULL)
+	{
+		fprintf (stderr, "Failed to load libmonosgen-2.0.dylib: %s\n", dlerror());
```
```suggestion
		fprintf(stderr, "Failed to load libmonosgen-2.0.dylib: %s\n", dlerror());
```

abcdefg30 commented on packaging/macos/apphost-mono.c:
```diff
@@ -0,0 +1,45 @@
+	if (getrlimit(RLIMIT_NOFILE, &limit) == 0 && limit.rlim_cur < 1024)
+	{
+		limit.rlim_cur = limit.rlim_max < 1024 ? limit.rlim_max : 1024;
+		setrlimit(RLIMIT_NOFILE, &limit);
```
This looks a bit odd out of context. Why do we need to limit (or increase the limit of) open file descriptors for the mono process?

pchote replied:
I don't actually know. This was originally copied from https://github.com/mono/monodevelop/blob/main/main/build/MacOSX/monostub.mm in our first native launcher and has been there ever since.

abcdefg30 replied:
Ok weird. I couldn't find any explanation on why that was added there either. Maybe we can add a comment here (saying where this comes from and that we don't know if it's actually necessary) and leave it as is?

pchote replied:
Tracked back to https://github.com/mono/monodevelop/commit/030ef63e0ccb1d3faac79ea24832fa286003ab13 which isn't very helpful.

abcdefg30 commented on packaging/macos/apphost.c:
```diff
@@ -0,0 +1,81 @@
+		return 1;
+	}
+
+	hostfxr_close_fn hostfxr_close = (hostfxr_close_fn)dlsym(lib, "hostfxr_close");
```
No check if `dlsym` returned `NULL`?

abcdefg30 commented on packaging/macos/apphost.c:
```diff
@@ -0,0 +1,81 @@
+	hostfxr_run_app(host_context_handle);
+
+	return hostfxr_close(host_context_handle);
+}
```
Missing newline.

abcdefg30 commented on packaging/macos/buildpackage.sh:
```diff
@@ -110,118 +107,139 @@ build_app() {
+modify_plist "{MINIMUM_SYSTEM_VERSION}" "10.11" "${TEMPLATE_DIR}/Contents/Info.plist"
+
+# Compile universal (x86_64 + arm64) Launcher and arch-specific apphosts
+clang apphost.c -o "${TEMPLATE_DIR}/Contents/MacOS/apphost-x86_64" -framework AppKit -target x86_64-apple-macos10.15
```
Do we need to check if `clang` is available or is this always installed by default on mac hosts.

abcdefg30 commented on packaging/macos/checkmono.c:
```diff
@@ -0,0 +1,106 @@
+		if (*req_version)
+			req_version++;
+
+		val = strtol (version, &end, 10);
```
```suggestion
		val = strtol(version, &end, 10);
```

abcdefg30 commented on packaging/macos/checkmono.c:
```diff
@@ -0,0 +1,106 @@
+
+typedef char *(* mono_get_runtime_build_info)(void);
+
+static int check_mono_version(const char *version, const char *req_version)
```
Why isn't this method used below but its contents duplicated there?

abcdefg30 commented on packaging/macos/launcher.m:
```diff
@@ -144,13 +199,45 @@ - (void)launchGameWithArgs: (NSArray *)gameArgs
+		size_t size;
+		cpu_type_t type;
+		size = sizeof(type);
+		sysctlbyname("hw.cputype", &type, &size, NULL, 0);
```
Do we need to check an error return code here?

abcdefg30 commented on packaging/macos/checkmono.c:
```diff
@@ -0,0 +1,106 @@
+		if (val > req_val)
+			return 0;
+
+		if (*req_version == '.' && *end != '.')
```
You already incremented `req_version` above, so the first conditional here is probably never true? If so we should move
```
		req_version = req_end;
		if (*req_version)
			req_version++;
```
below this check.

pchote replied:
Likewise here.

abcdefg30 reviewed (commented):
The first commit has "WIP" in its title. Do you want this PR to wait for the nuget packages to be ready?

pchote commented:
> The first commit has "WIP" in its title. Do you want this PR to wait for the nuget packages to be ready?

Correct (I mentioned this in the PR desc!)

abcdefg30 commented:
Ah derp, I went directly to the changes and then didn't read the OP afterwards. 🤦

PunkPun commented:
As predicted this build pretty much doubles the app size. Red Alert 106,8 MB (playtest) vs 206,9 MB (this devtest). Is this the right solution? These aren't large sizes but it still doesn't feel great

pchote commented:
IMO yes: the alternative is to provide three different dmg files and expect people to know which one to use. The size increase on the dmg is more modest (86 + 46 -> 103 MB) due to the deduplication and compression.

pchote committed:
Package macOS releases as a universal binary.

* Minimum macOS version is raised to 10.11.
* App bundles ship 3 versions of the runtime and engine binaries,
  and a fat launcher that selects the appropriate runtime/apphost.
* Mono is used for macOS 10.11 - 10.14, or if OPENRA_PREFER_MONO
  environment variable has been set.

pchote commented:
Updated the first commit with linux-arm64 deps to add to the nuget packages too.

pchote commented:
Updates the first commit and https://github.com/pchote/SDL2-binaries again to add the remaining platforms (linux-x64, windows-x86, windows-x64) so we ship the same versions for all.

Updated builds will be available from https://github.com/pchote/OpenRA/releases/tag/devtest-20221219. I currently only have access to my arm-based machine, so aren't able to test the Windows or linux-x64 builds!

Mailaender commented:
Can you submit a pull request towards https://github.com/OpenRA/SDL2-binaries?

Mailaender reviewed (approved):
linux-x64 still works

PunkPun commented:
Does this PR need a name change considering it also has changes to windows and linux?

Also are the nuget packages ready? I don't think we have much time to get this merged into prep

teinarss commented:
working github action for publishing sdl2 nuget https://github.com/teinarss/SDL2-CS/blob/master/.github/workflows/build-publish.yaml

pchote committed:
Update native deps to include macOS and Linux arm64 binaries.

pchote commented:
Updated nuget packages and added their sources to https://github.com/pchote/native-dependencies for now until we split them out into their own repos.

Updated this PR and created a new tag https://github.com/pchote/OpenRA/releases/tag/devtest-20221223

Ready to review, test, and hopefully merge.

PunkPun commented:
OSX 12.6.2 intel core runs the build
It also runs with argument OPENRA_PREFER_MONO=1

PunkPun commented:
Also runs on 13.0.1

Mailaender reviewed (changes requested):
AppImage does not work:

```ANGLE: SDL window creation failed: Could not initialize OpenGL / GLES library
Modern: SDL window creation failed: SDL not configured with OpenGL/GLX support
Embedded: SDL window creation failed: Could not initialize OpenGL / GLES library
System.InvalidOperationException: No supported OpenGL profiles were found.
   at OpenRA.Platforms.Default.Sdl2PlatformWindow..ctor(Size requestEffectiveWindowSize, WindowMode windowMode, Single scaleModifier, Int32 batchSize, Int32 videoDisplay, GLProfile requestProfile, Boolean enableLegacyGL) in /home/runner/work/OpenRA/OpenRA/OpenRA.Platforms.Default/Sdl2PlatformWindow.cs:line 164
   at OpenRA.Platforms.Default.DefaultPlatform.CreateWindow(Size size, WindowMode windowMode, Single scaleModifier, Int32 batchSize, Int32 videoDisplay, GLProfile profile, Boolean enableLegacyGL) in /home/runner/work/OpenRA/OpenRA/OpenRA.Platforms.Default/DefaultPlatform.cs:line 21
   at OpenRA.Renderer..ctor(IPlatform platform, GraphicSettings graphicSettings) in /home/runner/work/OpenRA/OpenRA/OpenRA.Game/Renderer.cs:line 79
   at OpenRA.Game.Initialize(Arguments args) in /home/runner/work/OpenRA/OpenRA/OpenRA.Game/Game.cs:line 385
ANGLE: SDL window creation failed: Could not initialize OpenGL / GLES library
Modern: SDL window creation failed: SDL not configured with OpenGL/GLX support
Embedded: SDL window creation failed: Could not initialize OpenGL / GLES library
System.InvalidOperationException: No supported OpenGL profiles were found.
   at OpenRA.Platforms.Default.Sdl2PlatformWindow..ctor(Size requestEffectiveWindowSize, WindowMode windowMode, Single scaleModifier, Int32 batchSize, Int32 videoDisplay, GLProfile requestProfile, Boolean enableLegacyGL) in /home/runner/work/OpenRA/OpenRA/OpenRA.Platforms.Default/Sdl2PlatformWindow.cs:line 164
   at OpenRA.Platforms.Default.DefaultPlatform.CreateWindow(Size size, WindowMode windowMode, Single scaleModifier, Int32 batchSize, Int32 videoDisplay, GLProfile profile, Boolean enableLegacyGL) in /home/runner/work/OpenRA/OpenRA/OpenRA.Platforms.Default/DefaultPlatform.cs:line 21
   at OpenRA.Renderer..ctor(IPlatform platform, GraphicSettings graphicSettings) in /home/runner/work/OpenRA/OpenRA/OpenRA.Game/Renderer.cs:line 79
   at OpenRA.Game.Initialize(Arguments args) in /home/runner/work/OpenRA/OpenRA/OpenRA.Game/Game.cs:line 385
```

pchote commented:
Updated.

PunkPun commented:
builds as well?

pchote commented:
https://github.com/pchote/OpenRA/releases/tag/devtest-20221223-2

pchote commented:
For completeness: The updates above scope creeped this PR to also supporting arm64 non-packaged development builds on both macOS and Linux.

pchote commented:
Reverted to the last known working freetype windows dlls.

https://github.com/pchote/OpenRA/releases/tag/devtest-20221223-3

PunkPun commented on .github/workflows/packaging.yml:
```diff
@@ -65,8 +65,8 @@ jobs:
-  macos-net:
-    name: macOS .NET
+  macos:
+    name: macOS
```
#20487 split the name `macOS Disk Images` into `macOS .NET` and `macOS Mono` Since we are getting back to a single GH action it might be better to revert to the original name

abcdefg30 reviewed (dismissed):
Confirmed that the x64 and x86 builds (devtest-20221223-3) work on my Windows machine. ✅

PunkPun commented:
[Changelogs](https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/15dd26321daa0c9c635331c04f2438ce542f4cc1...070ad685a2256a1846eb5f6dc8a57577a599defa)

https://github.com/OpenRA/OpenRA/compare/5e80fee913fb...f3f98d87507a
